### PR TITLE
node: add channel read/write type constraints

### PR DESCRIPTION
**Pull request by tbjump**
there are no functional changes in this PR. 

I am adding read/write type constraints to the main channels defined in cmd/guardiand/node.go. This makes the code easier to audit.

### CODE REVIEW

File: node/cmd/guardiand/node.go
Comment (jynnantonix) [inline node/cmd/guardiand/node.go:837]: "msg.EmiggerChain" -> "msg.EmitterChain"

File: node/cmd/guardiand/node.go
Comment (jynnantonix) [inline node/cmd/guardiand/node.go:846]: Should this be `logger.Fatal` rather than just an error?

File: node/cmd/guardiand/node.go
Comment (jynnantonix) [inline node/cmd/guardiand/node.go:841]: This will leak the goroutine, which isn't that big of an issue but if we want to do it the Right Way (TM) then we should select on the root context here:

```go
for {
	select {
		case <-rootCtx.Done():
			return
		case msg := <-c:
			// Handle msg
	}
}
```

**Review by jynnantonix — APPROVED** [review]
Overall idea looks good to me.

**Review by bruce-riley — APPROVED** [review]

File: node/cmd/guardiand/node.go
Comment (tbjump) [reply node/cmd/guardiand/node.go:801]: did we update the go version and can use generics now?

Comment (tbjump) [reply node/cmd/guardiand/node.go:801]: Looks like guardiand is still using go1.17? https://github.com/wormhole-foundation/wormhole/blob/f60e6669ec5e65ca2390a6ac10b834d2dec5821a/.github/workflows/build.yml#L254. Would love to see it upgraded though.

File: node/cmd/guardiand/node.go
Comment (tbjump) [reply node/cmd/guardiand/node.go:846]: thought about this as well. On the one hand, this is a failure that definitely needs to be looked at. On the other hand, other parts of the software could continue running fine. I think in the interest of catching bugs early during development, I can change this to logger.Fatal. What do you think?

File: node/cmd/guardiand/node.go
Comment (hendrikhofstadt) [reply node/cmd/guardiand/node.go:801]: Josh seems to be tackling it #1934

**Comment by tbjump** [comment]
refactored the channel pair creation using go generics. This requires go > 1.18 which is tackled in https://github.com/wormhole-foundation/wormhole/issues/1934. Will rebase once that's done but until then the tests for this PR will probably fail.

**Review by mtjmp — APPROVED** [review]

File: node/cmd/guardiand/node.go
Comment (kcsongor) [reply node/cmd/guardiand/node.go]: I really like the `channelPair` idea. On the type inference point, even if go will infer the types of those channels, I think we should still explicitly supply the type instantiation here. In my experience, relying on inference for return type polymorphic functions leads to subtle refactoring hazards and overall loss of readability.

**Review by jynnantonix — DISMISSED** [review]

**Review by jynnantonix — APPROVED** [review]

**Comment by hendrikhofstadt** [comment]
@tbjump pls rebase

**Comment by tbjump** [comment]
rebased and resolved conflicts. Also added `vaa.GetAllNetworkIDs()` which may be useful in other parts of the codebase. If there is an existing way to get all network IDs, that I did not find, please let me know!

**Comment by tbjump** [comment]
Codeowner review required: 
* @bruce-riley or @jynnantonix 
* @bruce-riley or @claudijd 
* @leoluk or @hendrikhofstadt 
* @jynnantonix or @hendrikhofstadt 
* @evan-gray or @jumpsiegel 
* @jumpsiegel 
* @evan-gray or @hendrikhofstadt 
* @bruce-riley or @hendrikhofstadt 
* @hendrikhofstadt 

Smallest sufficient set of reviewers: @hendrikhofstadt, @jumpsiegel, @bruce-riley

**Review by jumpsiegel — DISMISSED** [review]

File: sdk/vaa/structs.go
Comment (jynnantonix) [inline sdk/vaa/structs.go:296]: Can we keep this list in the same order as the consts below?  That will make it easier to see if any are missing.

**Review by jynnantonix — APPROVED** [review]

File: node/cmd/guardiand/node.go
Comment (hendrikhofstadt) [inline node/cmd/guardiand/node.go:857]: this could do a context aware write

**Review by hendrikhofstadt — DISMISSED** [review]

File: node/cmd/guardiand/node.go
Comment (bruce-riley) [inline node/cmd/guardiand/node.go:849]: I am not convinced that this check is worth the additional overhead and complexity, but I won't block the PR over it.

**Review by bruce-riley — APPROVED** [review]

**Review by hendrikhofstadt — APPROVED** [review]

**Review by hendrikhofstadt — DISMISSED** [review]

**Review by jynnantonix — DISMISSED** [review]
I'm assuming this PR is going to get squashed into a single commit.

**Comment by evan-gray** [comment]
@tbjump this has conflicts - converting to draft for now, but please address

File: sdk/vaa/structs.go
Comment (tbjump) [reply sdk/vaa/structs.go:307]: done

File: node/cmd/guardiand/node.go
Comment (tbjump) [reply node/cmd/guardiand/node.go:853]: can you explain / show me how?

**Comment by tbjump** [comment]
resolved merge conflicts.

**Review by hendrikhofstadt — APPROVED** [review]

**Review by jynnantonix — APPROVED** [review]
lgtm % Hendrik's comment

**Comment by tbjump** [comment]
@hendrikhofstadt @jumpsiegel @bruce-riley resolved all comments, resolved merge conflicts and passing all tests now. Please review again.

**Review by hendrikhofstadt — COMMENTED** [review]
Still has an open comment and needs a rebase

File: node/cmd/guardiand/node.go
Comment (tbjump) [reply node/cmd/guardiand/node.go]: are you ok with the latest version?

**Comment by tbjump** [comment]
resolved merge conflicts again

**Comment by bruce-riley** [comment]
While resolving merge conflicts in node.go, can you please change this to use the new approach?

**Comment by tbjump** [comment]
resolved merge conflicts again and addressed @bruce-riley 's todo. Please review once more.

**Review by bruce-riley — APPROVED** [review]

**Review by hendrikhofstadt — APPROVED** [review]

**Review by jumpsiegel — APPROVED** [review]

**Review by kcsongor — APPROVED** [review]

**Pull request merged**
